From bed7f57514db3a162988a4935f4c80965c53e3ad Mon Sep 17 00:00:00 2001
From: Ye Yang <yey@xilinx.com>
Date: Tue, 31 Aug 2021 15:30:54 +0800
Subject: [PATCH] add BGR format support

Signed-off-by: Ye Yang <yey@xilinx.com>
---
 sys/mediasrcbin/gstmediasrcbin.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/sys/mediasrcbin/gstmediasrcbin.c b/sys/mediasrcbin/gstmediasrcbin.c
index aa02aefac..e4f937228 100644
--- a/sys/mediasrcbin/gstmediasrcbin.c
+++ b/sys/mediasrcbin/gstmediasrcbin.c
@@ -278,6 +278,7 @@ get_media_bus_format (GstVideoFormat gst_fmt)
     case GST_VIDEO_FORMAT_NV16_10LE32:
       return MEDIA_BUS_FMT_UYVY10_1X20;
     case GST_VIDEO_FORMAT_RGB:
+    case GST_VIDEO_FORMAT_BGR:
       return MEDIA_BUS_FMT_RBG888_1X24;
     case GST_VIDEO_FORMAT_GRAY8:
       return MEDIA_BUS_FMT_Y8_1X8;
-- 
2.17.1

